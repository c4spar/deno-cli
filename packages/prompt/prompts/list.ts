import { encode } from 'https://deno.land/std@v0.41.0/encoding/utf8.ts';
import { blue, bold, dim, underline, yellow } from 'https://deno.land/std@v0.41.0/fmt/colors.ts';
import { stripeColors } from '../../table/lib/utils.ts';
import { Figures } from '../lib/figures.ts';
import { GenericInput, GenericInputPromptOptions, GenericInputPromptSettings } from './generic-input.ts';

export interface ListPromptOptions extends GenericInputPromptOptions<string[]> {
    separator?: string;
}

export interface ListPromptSettings extends GenericInputPromptSettings<string[]> {
    separator: string;
}

export class List extends GenericInput<string[], ListPromptSettings> {

    public static async prompt( options: ListPromptOptions ): Promise<string[] | undefined> {

        return new this( {
            pointer: blue( Figures.POINTER_SMALL ),
            separator: ',',
            ...options
        } ).run();
    }

    protected getMessage(): string {

        let message = ` ${ yellow( '?' ) } ${ bold( this.settings.message ) }`;

        if ( typeof this.settings.default !== 'undefined' ) {
            message += dim( ` (${ this.settings.default.join( `${ this.settings.separator } ` ) })` );
        }

        return message;
    }

    protected setMessage( message: string ) {

        message += ' ' + this.settings.pointer + ' ';

        const length = encode( stripeColors( message ) ).length;

        const oldInput: string = this.input;
        const oldInputParts: string[] = oldInput.trimLeft().split( this.regexp() );

        this.input = oldInputParts.join( `${ this.settings.separator } ` );

        message += oldInputParts.map( ( val: string ) => underline( val ) ).join( `${ this.settings.separator } ` );

        const inputDiff = oldInput.length - this.input.length;

        this.index -= inputDiff;

        this.write( message );

        this.screen.cursorTo( length - 1 + this.index );
    }

    protected regexp() {
        return new RegExp( this.settings.separator === ' ' ? ` +` : ` *${ this.settings.separator } *` );
    }

    protected addChar( char: string ): void {
        switch ( char ) {
            case this.settings.separator:
                if ( this.input.length && this.input.trim().slice( -1 ) !== this.settings.separator ) {
                    super.addChar( char );
                }
                break;
            default:
                super.addChar( char );
        }
    }

    protected deleteChar(): void {
        if ( this.input[ this.index - 1 ] === ' ' ) {
            super.deleteChar();
        }
        super.deleteChar();
    }

    protected sanitize( value: string ): string[] {
        return value.trim().split( this.regexp() );
    }

    protected validate( value: string[] | undefined ): boolean | string {
        return Array.isArray( value ) && ( value.every( ( val: string ) => val.trim().length > 0 ) || 'Value cannot be empty.' );
    }

    protected transform( value: string[] ): string {
        return value.join( `${ this.settings.separator } ` );
    }
}
