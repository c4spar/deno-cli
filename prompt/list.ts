import { stripeColors } from "../table/utils.ts";
import { blue, underline } from "./deps.ts";
import { Figures } from "./figures.ts";
import {
  GenericInput,
  GenericInputKeys,
  GenericInputPromptOptions,
  GenericInputPromptSettings,
} from "./_generic-input.ts";

export type ListKeys = GenericInputKeys;

export interface ListOptions extends GenericInputPromptOptions<string[]> {
  separator?: string;
  minLength?: number;
  maxLength?: number;
  minTags?: number;
  maxTags?: number;
  keys?: ListKeys;
}

interface ListSettings extends GenericInputPromptSettings<string[]> {
  separator: string;
  minLength: number;
  maxLength: number;
  minTags: number;
  maxTags: number;
  keys?: ListKeys;
}

export class List extends GenericInput<string[], ListSettings> {
  public static async prompt(options: string | ListOptions): Promise<string[]> {
    if (typeof options === "string") {
      options = { message: options };
    }

    return new this({
      pointer: blue(Figures.POINTER_SMALL),
      separator: ",",
      minLength: 0,
      maxLength: Infinity,
      minTags: 0,
      maxTags: Infinity,
      ...options,
    }).prompt();
  }

  protected setPrompt(message: string) {
    message += " " + this.settings.pointer + " ";

    const length = new TextEncoder().encode(stripeColors(message)).length;

    const oldInput: string = this.input;
    const oldInputParts: string[] = oldInput.trimLeft().split(this.regexp());

    this.input = oldInputParts.join(`${this.settings.separator} `);

    message += oldInputParts.map((val: string) => underline(val)).join(
      `${this.settings.separator} `,
    );

    const inputDiff = oldInput.length - this.input.length;

    this.index -= inputDiff;

    this.write(message);

    this.screen.cursorTo(length - 1 + this.index);
  }

  protected regexp() {
    return new RegExp(
      this.settings.separator === " " ? ` +` : ` *${this.settings.separator} *`,
    );
  }

  protected addChar(char: string): void {
    switch (char) {
      case this.settings.separator:
        if (
          this.input.length &&
          this.input.trim().slice(-1) !== this.settings.separator
        ) {
          super.addChar(char);
        }
        break;
      default:
        super.addChar(char);
    }
  }

  protected deleteChar(): void {
    if (this.input[this.index - 1] === " ") {
      super.deleteChar();
    }
    super.deleteChar();
  }

  protected validate(value: string): boolean | string {
    if (typeof value !== "string") {
      return false;
    }

    const values = this.transform(value).filter((val) => val !== "");

    for (const val of values) {
      if (val.length < this.settings.minLength) {
        return `Value must be longer then ${this.settings.minLength} but has a length of ${val.length}.`;
      }
      if (val.length > this.settings.maxLength) {
        return `Value can't be longer then ${this.settings.maxLength} but has a length of ${val.length}.`;
      }
    }

    if (values.length < this.settings.minTags) {
      return `The minimum number of tags is ${this.settings.minTags} but got ${values.length}.`;
    }
    if (values.length > this.settings.maxTags) {
      return `The maximum number of tags is ${this.settings.maxTags} but got ${values.length}.`;
    }

    return true;
  }

  protected transform(value: string): string[] {
    return value.trim().split(this.regexp());
  }

  protected format(value: string[]): string {
    return value.join(`, `);
  }
}
