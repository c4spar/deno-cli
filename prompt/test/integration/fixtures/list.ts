import { List } from "../../../list.ts";

export const tests = import.meta.main ? null : {
  "should enter some keywords": ["foo,bar", "\n"],
};

if (import.meta.main) {
  await List.prompt({
    message: "Enter some keywords",
    default: ["foo", "far"],
    hint: "some hint...",
  });
}
