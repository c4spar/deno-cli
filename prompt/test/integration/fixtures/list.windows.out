 ? Enter some keywords (foo, far) » 
 ❯ some hint...
\x1b[1A\x1b[1A\x1b[37G\x1b[G\x1b[0J ? Enter some keywords (foo, far) » bar, foo
 ❯ some hint...
\x1b[1A\x1b[1A\x1b[45G\x1b[G\x1b[0J ? Enter some keywords (foo, far) » bar, foo
\x1b[?25h\x1b[?25h